packages/kbn-config-schema/README.md

```diff
@@ -0,0 +1,511 @@
+
+__Output type:__ `number`
+
+__Options:__
+  * `defaultValue: number | Reference<number> | (() => number)` - defines a default value, see [Default values](#default-values) section for more details.
+  * `validate: (value: number) => string | void` - defines a custom validator function, see [Custom validation](#custom-validation) section for more details.
+  * `min: number` - defines a minimum value the number should have.
+  * `max: number` - defines a maximum value the number should have.
+
+__Usage:__
+```typescript
+const valueSchema = schema.number({ max: 10 });
+```
+
+__Notes:__
+* The `schema.number()` also supports a string as input if it can be safely coerced into number.
+
+#### `schema.boolean()`
+
+Validates input data as a boolean.
+
+__Output type:__ `boolean`
+
+__Options:__
+  * `defaultValue: boolean | Reference<boolean> | (() => boolean)` - defines a default value, see [Default values](#default-values) section for more details.
+  * `validate: (value: boolean) => string | void` - defines a custom validator function, see [Custom validation](#custom-validation) section for more details.
+
+__Usage:__
+```typescript
+const valueSchema = schema.boolean({ defaultValue: false });
+```
+
+#### `schema.literal()`
+
+Validates input data as a [string](https://www.typescriptlang.org/docs/handbook/advanced-types.html#string-literal-types), [numeric](https://www.typescriptlang.org/docs/handbook/advanced-types.html#numeric-literal-types) or boolean literal.
+
+__Output type:__ `string`, `number` or `boolean` literals
+
+__Options:__
+  * `defaultValue: TLiteral | Reference<TLiteral> | (() => TLiteral)` - defines a default value, see [Default values](#default-values) section for more details.
+  * `validate: (value: TLiteral) => string | void` - defines a custom validator function, see [Custom validation](#custom-validation) section for more details.
+
+__Usage:__
+```typescript
+const valueSchema = [
+  schema.literal('stringLiteral'),
+  schema.literal(100500),
+  schema.literal(false),
+];
+```
+
+### Composite types
+
+#### `schema.arrayOf()`
+
+Validates input data as a homogeneous array with the values being validated against predefined schema.
+
+__Output type:__ `TValue[]`
+
+__Options:__
+  * `defaultValue: TValue[] | Reference<TValue[]> | (() => TValue[])` - defines a default value, see [Default values](#default-values) section for more details.
+  * `validate: (value: TValue[]) => string | void` - defines a custom validator function, see [Custom validation](#custom-validation) section for more details.
+  * `minSize: number` - defines a minimum size the array should have.
+  * `maxSize: number` - defines a maximum size the array should have.
+
+__Usage:__
+```typescript
+const valueSchema = schema.arrayOf(schema.number());
+```
+
+#### `schema.object()`
+
+Validates input data as an object with a predefined set of properties.
+
+__Output type:__ `{ [K in keyof TProps]: TypeOf<TProps[K]> } as TObject`
```

Comment:
There are no references to `TProps` in the interface. BTW, if we had those types as comments in TSDoc it would solve the problem.
As a by-product:
- we can be sure that comments are always in sync with code.
- have suggestions in IDE.

Reply:
> There are no references to TProps in the interface.

True, I struggled to find a better way to describe properties types here. Do you have any suggestions?

> BTW, if we had those types as comments in TSDoc it would solve the problem

Yeah, but it sounds like a bigger separate effort, I'd leave it for a follow-up.